File: src/main/scala/com/vc/hard/L964.scala
package com.vc.hard

object L964 {
  def leastOpsExpressTarget(x: Int, target: Int): Int = {
    import scala.collection.mutable
    val map = new mutable.HashMap[Int, Int]()

    def solve(x: Int, target: Int): Int = {
      if(x == 1 && target == 1) return 0
      else if(target == 1) return 1
      else if(map.contains(target)) return map(target)
      else {
        var product = x.toLong
        var count = 0
        while(product < target) {
          product = product * x
          count += 1
        }

        //Target is multiple of x
        if(product.toInt == target) {
          map.put(target, count)
          return count
        }

        //case 1: product - something = target
        //i.e. something = product - target
        //we check product - target < target because we want next target to be less than
        //current target so that we can reach closer to 1 which is our exist condition
        val case1 = if(product - target < target.toLong)
          count + solve(x, (product - target).toInt) + 1
        else
          Int.MaxValue

        //case 2: product + something = target
        //i.e. something = target - product
        //we do product = product / x to make sure we don't have negative value after
        //we do target - product
        product = product / x
        val case2 = solve(x, target - product.toInt) + (if(count == 0) 2 else count)

        count = Math.min(case1, case2)
        map.put(target, count)
        count
      }
    }
    solve(x, target)
  }
}

File: src/main/java/com/vc/hard/L964.java
package com.vc.hard;

import java.util.HashMap;

class L964 {
    public int leastOpsExpressTarget(int x, int target) {
        return solve(x, target, new HashMap<Integer, Integer>());
    }

    private int solve(int x, int target, HashMap<Integer, Integer> memo) {
        if(x == target) return 0;
        else if(target == 1) return 1;
        else if(memo.containsKey(target)) return memo.get(target);

        //Do product *= x greedily to reach as much close as possible to target
        int count = 0;
        long product = x;
        while(product < target) {
            product *= x;
            count++;
        }

        if(product == target) return count;

        //Try subtracting 1 from product (x/x)
        int sub = 0;
        int nextTarget = (int)(product - target);
        if(nextTarget < target) sub = count + solve(x, nextTarget, memo) + 1;
        else sub = Integer.MAX_VALUE;

        //Go one step back(product /= x) and Try adding 1 to product (x/x)
        product /= x;
        nextTarget = (int)(target - product);
        int add = solve(x, nextTarget, memo) + (count == 0 ? 2 : count);

        int ans = Math.min(add, sub);

        memo.put(target, ans);
        return ans;
    }
}